### Задание 1
Скачать данные по ссылке https://www.kaggle.com/datasets/ionaskel/laptop-prices  
Считать данные с помощью pandas  
Вывести на экран первые 5 строк  
Посмотреть на описание признаков и на их содержание

In [48]:
import pandas as pd

In [49]:
df = pd.read_csv('kc-house-data.csv')
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

- id - Уникальный ID для каждого дома  
- date - Дата продажи дома  
- price - Стоимость продажи дома  
- bedrooms - Кол-во спален  
- bathrooms - Кол-во ванных комнат (0.5 - туалет без душа)  
- sqft_living - Кв. метры жилые  
- sqft_lot - Кв. метры общие  
- floors - Кол-во этажей  
- waterfront - Есть набержная или нет  
- view - Значение от 0 до 4 насколько хороший вид  
- condition - Значение от 1 до 5 насколько хорошее состояние  
- grade - Значение от 1 до 13, где 1-3 плохая конструкция здания и дизайн, 7 - средний уровень конструкции и дизайна, 11-13 - высокое качество конструкции и дизайна  
- sqft_above - Кв. метры дома, которые находятся выше земли  
- sqft_basement - Кв. метры дома, которые находятся ниже земли  
- yr_built - Год постройки дома  
- yr_renovated - Год ремонта дома  
- zipcode - Индекс  
- lat - Широтма  
- long - Долгота  
- sqft_living15 - Кв. метры жилой площади у 15 соседей  
- sqft_lot15 - Кв. метры общей площади у 15 соседей

### Задание 2
Проведите первичный анализ данных  
Изучите типы данных  
Найдите количество пропущенных ячеек в данных  
Посчитайте основные статистики по всем признакам и поизучайте их  
Пишите выводы!

In [50]:
df.shape #раземер датасета

(21613, 21)

In [51]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

В датасете присутствуют 3 типа данных: int, float, object

In [52]:
df.isnull().sum() #еще раз проверим на пропуски (пропусков нет)

id               0
date             0
price            0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
waterfront       0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
dtype: int64

In [53]:
df.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.161300e+04  2.161300e+04  21613.000000  21613.000000  21613.000000   
mean   4.580302e+09  5.400881e+05      3.370842      2.114757   2079.899736   
std    2.876566e+09  3.671272e+05      0.930062      0.770163    918.440897   
min    1.000102e+06  7.500000e+04      0.000000      0.000000    290.000000   
25%    2.123049e+09  3.219500e+05      3.000000      1.750000   1427.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    waterfront          view     condition  \
count  2.161300e+04  21613.000000  21613.000000  21613.000000  21613.000000   
mean   1.510697e+04      1.494309      0.007542      0.234303      3.409430   
std    4.142051e+04      0.539989      0.086517      0.766318      0.650743   
min    5.200000e+02      1.000000      0.000000      0.000000      1.000000   
25%    5.040000e+03      1.000000      0.000000      0.000000      3.000000   
50%    7.618000e+03      1.500000      0.000000      0.000000      3.000000   
75%    1.068800e+04      2.000000      0.000000      0.000000      4.000000   
max    1.651359e+06      3.500000      1.000000      4.000000      5.000000   

              grade    sqft_above  sqft_basement      yr_built  yr_renovated  \
count  21613.000000  21613.000000   21613.000000  21613.000000  21613.000000   
mean       7.656873   1788.390691     291.509045   1971.005136     84.402258   
std        1.175459    828.090978     442.575043     29.373411    401.679240   
min        1.000000    290.000000       0.000000   1900.000000      0.000000   
25%        7.000000   1190.000000       0.000000   1951.000000      0.000000   
50%        7.000000   1560.000000       0.000000   1975.000000      0.000000   
75%        8.000000   2210.000000     560.000000   1997.000000      0.000000   
max       13.000000   9410.000000    4820.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  
count  21613.000000  21613.000000  21613.000000   21613.000000   21613.000000  
mean   98077.939805     47.560053   -122.213896    1986.552492   12768.455652  
std       53.505026      0.138564      0.140828     685.391304   27304.179631  
min    98001.000000     47.155900   -122.519000     399.000000     651.000000  
25%    98033.000000     47.471000   -122.328000    1490.000000    5100.000000  
50%    98065.000000     47.571800   -122.230000    1840.000000    7620.000000  
75%    98118.000000     47.678000   -122.125000    2360.000000   10083.000000  
max    98199.000000     47.777600   -121.315000    6210.000000  871200.000000

### Задание 3
Ответьте на несколько вопросов  
#### 3.1 В каком диапазоне изменяются стоимости недвижимости?   

In [54]:
min_price = df['price'].min() # минимальная стоимость дома
min_price

75000.0

In [55]:
max_price = df['price'].max() # максимальная стоимость дома
max_price

7700000.0

In [56]:
median_price = df['price'].quantile(0.5) # медиана (средняя стоимость дома)
median_price

450000.0

In [57]:
mean_price = df['price'].mean() # средняя стоимость дома
mean_price

540088.1417665294

In [58]:
print(f'Стоимости недвижимости изменяются в диапазоне от {min_price} до {max_price}, средняя стоимость {mean_price.round(2)}.')

Стоимости недвижимости изменяются в диапазоне от 75000.0 до 7700000.0, средняя стоимость 540088.14.


#### 3.2 Какую долю в среднем занимают жилая площадь от всей площади по всем домам?  

Создадим новый признак доля жилой площади 

In [60]:
df['living_sq_share'] = df['sqft_living'] / df['sqft_lot']
df['living_sq_share'].head(10)

0    0.208850
1    0.354874
2    0.077000
3    0.392000
4    0.207921
5    0.053174
6    0.251503
7    0.109155
8    0.238286
9    0.288110
Name: living_sq_share, dtype: float64

In [62]:
df['living_sq_share'].mean()

0.3237447254204737

В среднем занимает жилая площадь от всей площади: 0.3237447254204737

#### 3.3 Как много домов с разными этажами в данных?  

In [73]:
df['floors'].value_counts()

1.0    10680
2.0     8241
1.5     1910
3.0      613
2.5      161
3.5        8
Name: floors, dtype: int64

In [75]:
len(df['floors'].unique()) # количество домов с разными этажами

6

#### 3.4 Насколько хорошие состояния у домов в данных?  

In [76]:
df['condition'].value_counts()

3    14031
4     5679
5     1701
2      172
1       30
Name: condition, dtype: int64

Наиболее часто встречаются дома в состоянии 3 (среднее), далее в состоянии 4 и 5

#### 3.5 Найдите года, когда построили первый дом, когда построили последний дом в данных?  

In [80]:
df['yr_built'].min() # год, когда построили первый дом

1900

In [81]:
df['yr_built'].max() # год, когда построили последний дом

2015

### Задание 4
Ответьте на несколько вопросов  
#### 4.1 Сколько в среднем стоят дома, у которых 2 спальни?  

In [84]:
df[df['bedrooms'] == 2]['price'].mean()

401372.681884058

#### 4.2 Какая в среднем общая площадь домов, у которых стоимость больше 600 000?  

In [87]:
df[df['price'] > 600000]['sqft_lot'].mean()

20442.524776214832

#### 4.3 Как много домов коснулся ремонт?  

In [97]:
df[df['yr_built'] <= df['yr_renovated']].count()

id                 914
date               914
price              914
bedrooms           914
bathrooms          914
sqft_living        914
sqft_lot           914
floors             914
waterfront         914
view               914
condition          914
grade              914
sqft_above         914
sqft_basement      914
yr_built           914
yr_renovated       914
zipcode            914
lat                914
long               914
sqft_living15      914
sqft_lot15         914
living_sq_share    914
dtype: int64

In [99]:
df[df['yr_renovated'] != 0].count() # 2-й способ

id                 914
date               914
price              914
bedrooms           914
bathrooms          914
sqft_living        914
sqft_lot           914
floors             914
waterfront         914
view               914
condition          914
grade              914
sqft_above         914
sqft_basement      914
yr_built           914
yr_renovated       914
zipcode            914
lat                914
long               914
sqft_living15      914
sqft_lot15         914
living_sq_share    914
dtype: int64

Ремонт коснулся 914 домов

#### 4.4 На сколько в среднем стоимость домов с оценкой grade домов выше 10 отличается от стоимости домов с оценкой grade меньше 4?

In [100]:
df[df['grade'] > 10]['price'].mean() - df[df['grade'] < 4]['price'].mean()

1488885.1175298805

#### 5.1 Выберите дом клиенту
Клиент хочет дом с видом на набережную, как минимум с тремя ванными и с подвалом. Сколько вариантов есть у клиента?

In [105]:
df[(df['waterfront'] == 1) & (df['bathrooms'] >= 3) & (df['sqft_basement'] > 0)]

id             date      price  bedrooms  bathrooms  \
300    3225069065  20140624T000000  3075000.0         4       5.00   
457    2122039094  20141126T000000   705000.0         3       3.00   
1164   1247600105  20141020T000000  5110800.0         5       5.25   
1261   7636800041  20140625T000000   995000.0         3       4.50   
1315   7558700030  20150413T000000  5300000.0         6       6.00   
1361   1925069082  20150511T000000  2200000.0         5       4.25   
2626   7738500731  20140815T000000  4500000.0         5       5.50   
2864   4114601570  20141118T000000  3600000.0         3       3.25   
3914   9808700762  20140611T000000  7062500.0         5       4.50   
3934   1269200229  20140723T000000  1379900.0         3       3.25   
5760    623059016  20140717T000000  1100000.0         4       3.25   
6233   2024069008  20140619T000000  2200000.0         5       4.75   
6508   4217402115  20150421T000000  3650000.0         6       4.75   
7989   9362000040  20140623T000000  3400000.0         3       4.50   
8092   1924059029  20140617T000000  4668000.0         5       6.75   
8163   1724069079  20150319T000000  1452000.0         2       3.25   
8784   3343302110  20150306T000000  1800000.0         3       3.00   
10263   625059036  20140813T000000  2700000.0         5       4.00   
11257  3761100045  20140618T000000  3000000.0         4       4.25   
11402  3426049284  20140819T000000  2300000.0         4       3.25   
11424  5119400075  20140620T000000   950000.0         3       3.25   
11446   624069035  20141209T000000  2750000.0         4       4.00   
11844  2623039082  20150218T000000   770000.0         3       3.50   
12459  4114601580  20140724T000000  1900000.0         6       4.00   
12935  3222049044  20140612T000000   835000.0         3       3.00   
13710  2923500230  20141216T000000  2600000.0         4       4.50   
13967  3126059023  20150303T000000  3395000.0         4       3.50   
14083  3625059043  20140904T000000  3300000.0         5       4.75   
14619  7936500109  20140725T000000  2230000.0         3       3.00   
14822  3623500408  20150330T000000  2600000.0         3       3.00   
15482   624069108  20140812T000000  3200000.0         4       3.25   
15668  4030100290  20141001T000000  1680000.0         5       3.50   
16817   825059178  20140923T000000  2574000.0         4       3.75   
17767  7352200450  20150115T000000  2050000.0         4       3.25   
18199  3625059152  20141230T000000  3300000.0         3       3.25   
18455  8043700300  20140608T000000  2700000.0         4       3.25   
18556  6613000930  20140902T000000  2950000.0         4       3.25   
20325   518500480  20140811T000000  3000000.0         3       3.50   
20767  8043700105  20150417T000000  2300000.0         4       4.00   
21201   518500460  20141008T000000  2230000.0         3       3.50   
21576  9253900271  20150107T000000  3567000.0         5       4.50   

       sqft_living  sqft_lot  floors  waterfront  view  ...  sqft_above  \
300           4550     18641     1.0           1     4  ...        2600   
457           1970     20978     2.0           1     3  ...        1770   
1164          8010     45517     2.0           1     4  ...        5990   
1261          4380     47044     2.0           1     3  ...        3720   
1315          7390     24829     2.0           1     4  ...        5000   
1361          4640     22703     2.0           1     4  ...        2860   
2626          6640     40014     2.0           1     4  ...        6350   
2864          5020     12431     2.0           1     4  ...        3420   
3914         10040     37325     2.0           1     2  ...        7680   
3934          3786     38038     1.0           1     4  ...        1934   
5760          3190     11774     2.0           1     4  ...        2610   
6233          5990     10450     2.0           1     4  ...        4050   
6508          5480     19401     1.5           1     4  ...        3910   
7989          5230     1

41 вариант

#### 5.2 Выберите дом клиенту
Клиент хочет дом либо с очень красивым видом из окна, либо с видом на набережную, в очень хорошем состоянии и год постройки не меньше 1980 года. В какой ценовом диапазоне будут дома?

In [109]:
df[((df['view'] == 4) | (df['waterfront'] == 1)) & (df['condition'] == 5) & (df['yr_built']) >= 1980]

Empty DataFrame
Columns: [id, date, price, bedrooms, bathrooms, sqft_living, sqft_lot, floors, waterfront, view, condition, grade, sqft_above, sqft_basement, yr_built, yr_renovated, zipcode, lat, long, sqft_living15, sqft_lot15, living_sq_share]
Index: []

[0 rows x 22 columns]

Таких домов нет (если я правильно поняла задание). Второй вариант ниже.

In [116]:
df[(df['view'] == 4) | ((df['waterfront'] == 1) & (df['condition'] == 5) & (df['yr_built']) >= 1980)]

id             date      price  bedrooms  bathrooms  \
21     2524049179  20140826T000000  2000000.0         3       2.75   
58     7922800400  20140827T000000   951000.0         5       3.25   
230    8096000060  20150413T000000   655000.0         2       1.75   
246    2025069065  20140929T000000  2400000.0         4       2.50   
264    2123039032  20141027T000000   369900.0         1       0.75   
...           ...              ...        ...       ...        ...   
20767  8043700105  20150417T000000  2300000.0         4       4.00   
20923  7852070210  20140527T000000  1149000.0         4       3.00   
21050  2424059170  20150219T000000   900000.0         5       6.00   
21201   518500460  20141008T000000  2230000.0         3       3.50   
21576  9253900271  20150107T000000  3567000.0         5       4.50   

       sqft_living  sqft_lot  floors  waterfront  view  ...  sqft_above  \
21            3050     44867     1.0           0     4  ...        2330   
58            3250     14342     2.0           0     4  ...        3250   
230           1450     15798     2.0           1     4  ...        1230   
246           3650      8354     1.0           1     4  ...        1830   
264            760     10079     1.0           1     4  ...         760   
...            ...       ...     ...         ...   ...  ...         ...   
20767         4360      8175     2.5           1     4  ...        3940   
20923         5940     11533     2.0           0     4  ...        4950   
21050         7120     40806     2.0           0     4  ...        5480   
21201         3760      5634     2.0           1     4  ...        2830   
21576         4850     10584     2.0           1     4  ...        3540   

       sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
21               720      1968             0    98040  47.5316 -122.233   
58                 0      1968             0    98008  47.5880 -122.116   
230              220      1915          1978    98166  47.4497 -122.375   
246             1820      2000             0    98074  47.6338 -122.072   
264                0      1936             0    98070  47.4683 -122.438   
...              ...       ...           ...      ...      ...      ...   
20767            420      2007             0    98008  47.5724 -122.104   
20923            990      2004             0    98065  47.5443 -121.870   
21050           1640      2007             0    98006  47.5451 -122.114   
21201            930      2014             0    98056  47.5285 -122.205   
21576           1310      2007             0    98008  47.5943 -122.110   

       sqft_living15  sqft_lot15  living_sq_share  
21              4110       20336         0.067979  
58              2960       11044         0.226607  
230             2030       13193         0.091784  
246             3120       18841         0.436916  
264             1230       14267         0.075404  
...              ...         ...              ...  
20767           2670        8525         0.533333  
20923           4240       12813         0.515044  
21050           3440       36859         0.174484  
21201           3560        5762         0.667377  
21576           3470       18270         0.458239  

[319 rows x 22 columns]

Минимальный прайс

In [120]:
df[(df['view'] == 4) | ((df['waterfront'] == 1) & (df['condition'] == 5) & (df['yr_built']) >= 1980)]['price'].min()

252000.0

Максимальный прайс

In [121]:
df[(df['view'] == 4) | ((df['waterfront'] == 1) & (df['condition'] == 5) & (df['yr_built']) >= 1980)]['price'].max()

6885000.0

#### 5.3 Выберите дом клиенту
Клиент хочет дом без подвала, с двумя этажами, стоимостью до 150000. Какая оценка по состоянию у таких домов в среднем?

In [114]:
df[(df['sqft_basement'] == 0) & (df['floors'] == 2) & (df['price'] < 150000)]

id             date     price  bedrooms  bathrooms  \
2589   5061300030  20140508T000000  134000.0         2       1.50   
10105  5466310060  20150324T000000  139500.0         2       1.50   
15610  7614100080  20150211T000000  140000.0         3       1.75   
18332  9272202260  20140924T000000  130000.0         3       1.00   
18939  4239400300  20141129T000000   90000.0         3       1.00   
18988  7399100210  20141126T000000  140000.0         3       1.50   

       sqft_living  sqft_lot  floors  waterfront  view  ...  sqft_above  \
2589           980      5000     2.0           0     0  ...         980   
10105         1230      1561     2.0           0     0  ...        1230   
15610         1270      8991     2.0           0     0  ...        1270   
18332         1200      7000     2.0           0     0  ...        1200   
18939          980      2490     2.0           0     0  ...         980   
18988         1200      2002     2.0           0     0  ...        1200   

       sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
2589               0      1922          2003    98014  47.7076 -121.359   
10105              0      1983             0    98042  47.3768 -122.149   
15610              0      1981             0    98042  47.3563 -122.149   
18332              0      1908             0    98116  47.5883 -122.384   
18939              0      1969             0    98092  47.3170 -122.182   
18988              0      1966             0    98055  47.4659 -122.189   

       sqft_living15  sqft_lot15  living_sq_share  
2589            1040        5000         0.196000  
10105           1660        2243         0.787956  
15610           1270        8993         0.141252  
18332           3290        6000         0.171429  
18939            980        3154         0.393574  
18988           1270        1848         0.599401  

[6 rows x 22 columns]

Признак категориальный: лучше подойдет медиана

In [115]:
df[(df['sqft_basement'] == 0) & (df['floors'] == 2) & (df['price'] < 150000)]['condition'].median()

3.0